src/source-map.ts

```diff
@@ -0,0 +1,192 @@
+      newMapping.name = mapping.name;
+    }
+
+    generator.addMapping(newMapping);
+  });
+
+  return generator.toJSON();
+}
+
+export async function getExistingSourcemap(
+    analyzer: AnalysisContext, sourceUrl: string, sourceContent: string) {
+  const dataUrlPrefix = 'data:application/json;';
+  const sourceMappingUrlParts = sourceMappingUrlExpr.exec(sourceContent);
+  if (sourceMappingUrlParts === null) {
+    return null;
+  }
+
+  let sourcemap: RawSourceMap;
+  let mapUrl = sourceUrl;
+  if (sourceMappingUrlParts[1].substr(0, dataUrlPrefix.length) ===
+      dataUrlPrefix) {
+    sourcemap = JSON.parse(Buffer.from(sourceMappingUrlParts[1], 'base64')
+                               .toString('utf8')) as RawSourceMap;
+  } else {
+    mapUrl = urlLib.resolve(sourceUrl, sourceMappingUrlParts[1].trim());
+    sourcemap = JSON.parse(await analyzer.load(mapUrl)) as RawSourceMap;
+  }
+
+  // Rewrite the sources array to be relative to the current URL
+  if (sourcemap.sources) {
+    sourcemap.sources =
+        sourcemap.sources.map(source => urlLib.resolve(mapUrl, source));
+  }
+  return sourcemap;
+}
+
+export async function addOrUpdateSourcemapComment(
+    analyzer: AnalysisContext,
+    sourceUrl: string,
+    sourceContent: string,
+    lineOffset: number = 0,
+    firstLineCharOffset: number = 0) {
+  let sourcemap =
+      await getExistingSourcemap(analyzer, sourceUrl, sourceContent);
+
+  let hasExisting = true;
+  if (sourcemap === null) {
+    hasExisting = false;
+    sourcemap = createJsIdentitySourcemap(
+        sourceUrl, sourceContent, lineOffset, firstLineCharOffset);
+  }
+
+  if (sourcemap === null) {
+    return sourceContent;
+  }
+
+  let updatedSourcemapComment = inlineSourcemapPrefix +
+      Buffer.from(JSON.stringify(sourcemap), 'utf8').toString('base64') + '\n';
+  if (hasExisting) {
+    return sourceContent.replace(sourceMappingUrlExpr, updatedSourcemapComment);
+  } else {
+    if (sourceContent.length > 0 &&
+        sourceContent[sourceContent.length - 1] === '\n') {
+      updatedSourcemapComment = updatedSourcemapComment.substr(1);
+    }
+
+    return sourceContent + updatedSourcemapComment;
+  }
+}
+
+
+/**
+ * Update the source maps in inline script elements to reflect their relative
+ * line offset in the bundle.
+ */
+export function updateSourcemapLocations(ast: parse5.ASTNode) {
+  // We need to serialize and reparse the dom for updated location information
+  ast = parse5.parse(parse5.serialize(ast), {locationInfo: true});
+  const inlineScripts = dom5.queryAll(ast, matchers.inlineJavascript);
+  inlineScripts.forEach(script => {
+    let content = dom5.getTextContent(script);
+
+    const sourceMapContentParts = sourceMapCommentExpr.exec(content);
+    if (!sourceMapContentParts) {
+      return;
+    }
+    const scriptLocation =
+        (script.__location as parse5.ElementLocationInfo).startTag;
+    const sourceMapContent = Buffer.from(sourceMapContentParts[2], 'base64');
```

Comment:
This logic is repeated a couple places, but it's somewhat noisy. Mind extracting it to a couple of module-local functions? i.e. two functions `(map: RawSourceMap) => string` and `(encodedMap: string) => RawSourceMap`.